File: algos/singlyLinkedLists/w2d5_remove_neg_2nd_to_last_partition.py
import random

class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

class Singly_Linked_List:
    def __init__(self):
        self.head = None

    def is_empty(self):
        return self.head == None

    def populate_random(self, max, length):
        for i in range(length):
            temp = random.randrange(1, max)
            self.add_front(temp)
        return self

    def add_front(self, val):
        new_head = Node(val)
        new_head.next = self.head
        self.head = new_head
        return self

    def contains(self, val):
        if (self.is_empty()):
            print("This SLL is empty.")
            return self
        else:
            runner = self.head
            while (runner):
                if (runner.val == val):
                    print(f'Value ({val}) found.')
                    return True
                runner = runner.next
            print(f'Value ({val}) not found.')
            return False

    def print_vals(self):
        if (self.is_empty()):
            print("This SLL is empty.")
            return self
        else:
            runner = self.head
            output = '**********\n'
            while (runner):
                output += f'{str(runner.val)} > '
                runner = runner.next
            print(output)
        return self

    # SLList: Remove Negatives
    # Create a function that removes all negative values from a list, then return the modified list.
    def remove_negatives(self):
        prev = None
        runner = self.head

        while (runner):
            if(runner.val < 0):
                if(prev == None):
                    self.head = runner.next
                else:
                    prev.next = runner.next
            prev = runner
            runner = runner.next
        return self
    
    # Second to Last Value
    # Create a function that returns the value in the second-to-last node.
    def second_to_last_val(self):
        runner = self.head
        if(runner == None or runner.next == None):
            print("This list does not contain enough nodes.")
            return self
        else:
            prev = None
            while(runner.next):
                prev = runner
                runner = runner.next
            print(f'Second-to-last val: {prev.val}')
        return self
    
    # Partition
    # Create a function that, given a value, locates the first node with that value, moves all nodes with values less than that value to be earlier, and moves all nodes with values greater than that value to be later. Otherwise, original order need not be perfectly preserved. Return the list when complete.
    def partition(self, val):
        if (self.is_empty()):
            print("This SLL is empty.")
            return self
        else:
            if(not self.contains(val)):
                return self
            else:
                lowerList = Singly_Linked_List()
                higherList = Singly_Linked_List()
                valCount = 0
                runner = self.head

                while(runner):
                    if (runner.val < val):
                        lowerList.add_front(runner.val)
                    elif (runner.val > val):
                        higherList.add_front(runner.val)
                    else:
                        valCount += 1
                    runner = runner.next
                for i in range(valCount):
                    higherList.add_front(val)
                runner = lowerList.head
                while(runner.next):
                    runner = runner.next
                runner.next = higherList.head
                self.head = lowerList.head
        return self

mySLL = Singly_Linked_List()
mySLL.populate_random(300, 10).add_front(150).populate_random(300, 10).print_vals()
mySLL.partition(150).print_vals()

File: algos/singlyLinkedLists/w2d5_removeNeg_2ndToLast_Partition.js
class Node {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

class SinglyLinkedList {
    constructor() {
        this.head = null;
    }

    isEmpty() {
        return this.head === null;
    }

    populateRandom(max, length) {
        for (let i = 1; i <= length; i++) {
            let temp = Math.floor(Math.random() * max) + 1;
            this.addFront(temp);
        }
        return this;
    }

    addFront(val) {
        const newHead = new Node(val);
        newHead.next = this.head;
        this.head = newHead;
        return this;
    }

    contains(val) {
        if (this.isEmpty()) {
            console.log("This SLL is empty.")
            return this;
        } else {
            let runner = this.head;
            while (runner) {
                if (runner.val === val) {
                    console.log(`Value (${val}) found.`)
                    return true;
                }
                runner = runner.next;
            }
            console.log(`Value (${val}) not found.`)
            return false;
        }
    }

    printVals() {
        if (this.isEmpty()) {
            console.log("This SLL is empty.")
            return this;
        } else {
            let runner = this.head;
            let output = `**********\n`
            while (runner) {
                output += `${runner.val.toString()} > `;
                runner = runner.next;
            }
            console.log(output);
        }
        return this;
    }

    // SLList: Remove Negatives
    // Create a function that removes all negative values from a list, then return the modified list.
    removeNegatives() {
        let prev = null;
        let runner = this.head;

        while (runner) {
            if(runner.val < 0) {
                if(prev === null) {
                    this.head = runner.next;
                } else {
                    prev.next = runner.next;
                }
            }
            prev = runner;
            runner = runner.next;
        }
        return this;
    }
    
    // Second to Last Value
    // Create a function that returns the value in the second-to-last node.
    secondToLastVal() {
        let runner = this.head;
        if(runner === null || runner.next == null) {
            console.log("This list does not contain enough nodes.")
            return this;
        } else {
            let prev = null;
            while(runner.next) {
                prev = runner;
                runner = runner.next;
            }
            console.log(`Second-to-last val: ${prev.val}`)
        }
        return this;
    }
    
    // Partition
    // Create a function that, given a value, locates the first node with that value, moves all nodes with values less than that value to be earlier, and moves all nodes with values greater than that value to be later. Otherwise, original order need not be perfectly preserved. Return the list when complete.
    partition(val) {
        if (this.isEmpty()) {
            console.log("This SLL is empty.")
            return this;
        } else {
            if(!this.contains(val)) {
                return this;
            } else {
                let lowerList = new SinglyLinkedList();
                let higherList = new SinglyLinkedList();
                let valCount = 0;
                let runner = this.head;

                while(runner) {
                    if (runner.val < val) {
                        lowerList.addFront(runner.val);
                    } else if (runner.val > val) {
                        higherList.addFront(runner.val)
                    } else {
                        valCount++;
                    }
                    runner = runner.next
                }
                for (let i = 0; i < valCount; i++) {
                    higherList.addFront(val);
                }
                runner = lowerList.head;
                while(runner.next){
                    runner = runner.next;
                }
                runner.next = higherList.head;
                this.head = lowerList.head;
            }
        }
        return this;
    }
}